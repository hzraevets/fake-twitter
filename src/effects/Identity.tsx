import React, {
  Dispatch,
  SetStateAction,
  useEffect,
  createContext,
  useState,
  ReactNode,
} from 'react';

import { UserStorage, User } from 'models';
import { md5 } from 'utils/md5';

export interface IdentityContextInterface {
  identity: UserStorage;
  setIdentity: Dispatch<SetStateAction<UserStorage>>;
  loginUser: string | null;
  setLoginUser: Dispatch<SetStateAction<string | null>>;
  logout: () => void;
  register: (newUser: User) => UserStorage;
}

const IdentityContext = createContext<IdentityContextInterface>({
  identity: {},
  setIdentity: () => null,
  loginUser: null,
  setLoginUser: () => null,
  logout: () => null,
  register: () => ({}),
});

const getIdentity = (): UserStorage => {
  const identity = localStorage.getItem('identity');
  if (!identity) {
    localStorage.setItem('identity', '{}');
    return {};
  }

  return JSON.parse(identity);
};

const getLoginUser = (): string | null => {
  const loginUser = localStorage.getItem('loginUser');
  if (!loginUser) {
    return null;
  }

  return loginUser;
};

const IdentityProvider = ({ children }: { children: ReactNode }) => {
  const [ identity, setIdentity ] = useState(getIdentity);
  const [ loginUser, setLoginUser ] = useState(getLoginUser);

  function logout() {
    if (loginUser) {
      setLoginUser(null);
      localStorage.removeItem('loginUser');
    }
  }

  function register(newUser: User) {
    const newStorage: UserStorage = {
      ...identity,
      [newUser.username]: {
        ...newUser,
        password: md5(newUser.password),
      },
    };

    localStorage.setItem('identity', JSON.stringify(newStorage));
    return newStorage;
  }

  useEffect(() => {
    const updateIdentity = () => {
      localStorage.setItem('identity', JSON.stringify(identity));
    };

    updateIdentity();
  }, [identity]);

  useEffect(() => {
    const executeLogin = () => {
      if (loginUser) {
        localStorage.setItem('loginUser', loginUser);
      }
    };

    executeLogin();
  }, [loginUser]);

  return (
    <IdentityContext.Provider
      value={{
        identity,
        setIdentity,
        loginUser,
        setLoginUser,
        logout,
        register,
      }}
    >
      {children}
    </IdentityContext.Provider>
  );
};

export { IdentityContext, IdentityProvider };
