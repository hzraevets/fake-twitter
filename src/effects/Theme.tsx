import React, {
  Dispatch,
  SetStateAction,
  useEffect,
  useCallback,
  createContext,
  useState,
  ReactNode,
} from 'react';
import { ConfigProvider, theme as antdTheme } from "antd";

interface ThemeContextInterface {
  theme: string;
  setTheme: Dispatch<SetStateAction<string>>;
  toggleTheme: () => void;
}

const ThemeContext = createContext<ThemeContextInterface>({
  theme: 'light',
  setTheme: () => null,
  toggleTheme: () => null,
});

const getTheme = () => {
  const theme = localStorage.getItem('theme');
  if (!theme) { // default theme
    localStorage.setItem('theme', 'light');
    return 'light';
  }

  return theme;
};

const ThemeProvider = ({ children }: { children: ReactNode }) => {
  const [ theme, setTheme ] = useState(getTheme);

  function toggleTheme() {
    if (theme === 'dark') {
      setTheme('light');
    } else {
      setTheme('dark');
    }
  }

  const refreshTheme = useCallback(() => {
    localStorage.setItem('theme', theme);
  }, [theme]);

  useEffect(() => {
    refreshTheme();
  }, [refreshTheme, theme]);

  return (
    <ThemeContext.Provider
      value={{
        theme,
        setTheme,
        toggleTheme,
      }}
    >
      <div className={`theme-container ${theme}`}>
        <ConfigProvider
          theme={{
            algorithm: theme === 'dark'
              ? antdTheme.darkAlgorithm
              : antdTheme.defaultAlgorithm
          }}
        >
          {children}
        </ConfigProvider>
      </div>
    </ThemeContext.Provider>
  );
};

export { ThemeContext, ThemeProvider };
