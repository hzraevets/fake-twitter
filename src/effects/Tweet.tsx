import React, { createContext, useState, ReactNode } from 'react';
import getUnixTime from 'date-fns/getUnixTime';

import {
  Tweet,
  TweetStorage,
  CreateNewTweetParams,
  ReadNonDeleteTweetParams,
  ReadNonDeleteTweetResponse,
} from 'models';

export interface TweetContextInterface {
  allTweets: TweetStorage;
  nextTweetId: number;
  addNewTweet: (params: CreateNewTweetParams) => void;
  readTimeline: (params: ReadNonDeleteTweetParams) => Promise<ReadNonDeleteTweetResponse>;
  editTweet: (id: number, content: string) => void;
  deleteTweet: (id: number) => void;
}

const initialTweetStorage: TweetStorage = {
  hashMap: {},
  order: [],
};

const TweetContext = createContext<TweetContextInterface>({
  allTweets: initialTweetStorage,
  nextTweetId: 1,
  addNewTweet: () => null,
  readTimeline: () => Promise.resolve({ data: [] }),
  editTweet: () => null,
  deleteTweet: () => null,
});

const TweetProvider = ({ children }: { children: ReactNode }) => {
  const [ allTweets, setAllTweets ] = useState(initialTweetStorage);
  const [ nextTweetId, setNextTweetId ] = useState(1);

  function addNewTweet(params: CreateNewTweetParams) {
    const now = getUnixTime(new Date());
    const { hashMap, order } = allTweets;

    const newTweet: Tweet = {
      ...params,
      id: nextTweetId,
      createTime: now,
      updateTime: now,
      isDeleted: false,
    };

    setNextTweetId(nextTweetId + 1);
    setAllTweets({
      hashMap: { ...hashMap, [nextTweetId]: newTweet },
      order: [ nextTweetId, ...order ],
    });
  }

  function readTimeline({ start, offset = 1 }: ReadNonDeleteTweetParams) {
    const data: Tweet[] = [];
    let nextCursor: number | null;

    const { hashMap, order } = allTweets;

    for (nextCursor = start; (data.length < offset) && nextCursor < order.length; nextCursor++) {
      const tweet = hashMap[order[nextCursor]];

      if (!tweet.isDeleted) {
        data.push(tweet);
      }
    }

    const res: ReadNonDeleteTweetResponse = { data };

    if (nextCursor < order.length) {
      res.nextCursor = nextCursor;
    }

    return new Promise<ReadNonDeleteTweetResponse>(resolve =>
      setTimeout(() => resolve(res), Math.random() * 1000));
  }

  function editTweet(id: number, content: string) {
    const { hashMap, order } = allTweets;
    const oldTweet = hashMap[id];

    if (!oldTweet) {
      // This should not happen!
      return;
    }

    hashMap[id].content = content;
    hashMap[id].updateTime = getUnixTime(new Date());

    setAllTweets({ hashMap, order });
  }

  function deleteTweet(id: number) {
    const { hashMap } = allTweets;
    let order = allTweets.order;
    const oldTweet = hashMap[id];

    if (!oldTweet) {
      // This should not happen!
      return;
    }

    hashMap[id].isDeleted = true;
    order = order.filter(tid => tid !== id);

    setAllTweets({ hashMap, order });
  }

  return (
    <TweetContext.Provider
      value={{
        allTweets,
        nextTweetId,
        addNewTweet,
        readTimeline,
        editTweet,
        deleteTweet,
      }}
    >
      {children}
    </TweetContext.Provider>
  );
}

export { TweetContext, TweetProvider };
