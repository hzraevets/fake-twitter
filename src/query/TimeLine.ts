import { InfiniteData, QueryClient } from 'react-query';
import getUnixTime from 'date-fns/getUnixTime';

import {
  CreateNewTweetParams,
  ReadNonDeleteTweetParams,
  ReadNonDeleteTweetResponse,
  Tweet,
} from 'models';

export const timeLineQuery = new QueryClient();

export const getNextPageParam = (lastPage: ReadNonDeleteTweetResponse, pages: ReadNonDeleteTweetResponse[]) => {
  return lastPage.nextCursor;
}

export const generateFetchTimeline = (
  readTimeline:  (params: ReadNonDeleteTweetParams) => Promise<ReadNonDeleteTweetResponse>
) => ({ pageParam = 0 }) => {
  return readTimeline({ start: pageParam });
}

export const reFetchQuery = async (
  readTimeline:  (params: ReadNonDeleteTweetParams) => Promise<ReadNonDeleteTweetResponse>,
  clearQuery: boolean = false,
) => {
  if (clearQuery) {
    timeLineQuery.clear();
  }

  await setTimeout(() => null, 100);

  return timeLineQuery.prefetchInfiniteQuery('timeline', generateFetchTimeline(readTimeline), {
    getNextPageParam,
  });
}
export function flattenQueryPages(pages: ReadNonDeleteTweetResponse[]) {
  const flattenTweetList: Tweet[] = [];

  pages.forEach((group: ReadNonDeleteTweetResponse) => {
    flattenTweetList.push(...group.data);
  });

  return flattenTweetList;
}

export function addNewTweetMutation({ author, firstname, content, addNewTweet }: {
  author: string;
  firstname: string;
  content: string;
  addNewTweet: (params: CreateNewTweetParams) => Tweet | null;
}) {
  const now = getUnixTime(new Date());

  const newTweet = addNewTweet({
    author,
    firstname,
    content,
  });

  if (!newTweet) {
    return Promise.resolve({
      id: -1,
      author,
      firstname,
      content,
      createTime: now,
      updateTime: now,
      isDeleted: false,
    });
  }

  return Promise.resolve(newTweet);
}

export function onAddNewTweetMutationSuccess(newTweet: Tweet) {
  timeLineQuery.setQueryData<InfiniteData<ReadNonDeleteTweetResponse>>('timeline', (oldData: InfiniteData<ReadNonDeleteTweetResponse> | undefined) => {
    const nextData: InfiniteData<ReadNonDeleteTweetResponse> = {
      pages: [],
      pageParams: [],
    };

    if (oldData) {
      nextData.pages = [ ...oldData.pages ];
      nextData.pageParams = [ ...oldData.pageParams ];
    }

    if (newTweet) {
      nextData.pages.unshift({
        data: [ newTweet ],
      });
      nextData.pageParams.unshift(undefined);
    }

    return nextData;
  });
}
