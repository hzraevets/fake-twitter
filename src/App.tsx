import React, { useContext } from "react";
import { RouterProvider } from 'react-router-dom';

import { generateAppRoutes } from './index-routing';
import { IdentityContext } from 'effects/Identity';

export function App() {
  const { loginUser } = useContext(IdentityContext);

  return (
    <div className="app relative h-screen w-full min-w-min bg-blue-400 dark:bg-black text-black dark:text-white">
      <RouterProvider router={generateAppRoutes(loginUser)} />
    </div>
  );
}
