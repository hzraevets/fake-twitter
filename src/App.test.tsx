import React from 'react';
import { act } from 'react-dom/test-utils';
import { render } from '@testing-library/react';
import { QueryClientProvider } from 'react-query';
// @ts-ignore
import { create } from 'react-test-renderer';

import { App } from './App';
import { ThemeProvider } from 'effects/Theme';
import { IdentityProvider } from 'effects/Identity';
import { TweetProvider } from 'effects/Tweet';
import { timeLineQuery } from 'query/TimeLine';

describe('App', () => {
  it('render App correctly', async () => {
    // eslint-disable-next-line testing-library/no-unnecessary-act
    await act(async () => {
      render(
        <TweetProvider>
          <ThemeProvider>
            <IdentityProvider>
              <QueryClientProvider client={timeLineQuery}>
                <App/>
              </QueryClientProvider>
            </IdentityProvider>
          </ThemeProvider>
        </TweetProvider>
      );
    });
  });

  it('render App to snapshot', () => {
    const renderer = create(<App/>);
    expect(renderer.toJSON()).toMatchSnapshot();
  });
});
