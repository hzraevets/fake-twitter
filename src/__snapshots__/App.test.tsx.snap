// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App render App to snapshot 1`] = `
<div
  className="app relative h-screen w-full min-w-min bg-blue-400 dark:bg-black text-black dark:text-white"
/>
`;
